File: Algorithm/python/2751_수정렬하기2.py
import sys

sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline


def quickSort(left, right):
    if left >= right:
        return
    l = left
    r = right
    mid = arr[(l + r) // 2]
    while True:
        while arr[l] < mid:
            l += 1
        while arr[r] > mid:
            r -= 1
        if l >= r:
            break
        temp = arr[l]
        arr[l] = arr[r]
        arr[r] = temp
    quickSort(left, l - 1)
    quickSort(r + 1, right)


n = int(input().strip())
arr = list()
for i in range(n):
    arr.append(int(input().strip()))
quickSort(0, n - 1)
print('\n'.join(str(x) for x in arr))

File: Algorithm/baekjoon/c++/2751_수정렬하기2.cpp
#include<cstdio>
int *arr;

void quicksort(int left, int right) {
	if (left >= right)
		return;
	int l = left - 1;
	int r = right + 1;
	int mid = arr[(l + r) / 2];
	while (true) {
		while (arr[++l] < mid);
		while (arr[--r] > mid);
		if (l >= r)
			break;
		int temp = arr[l];
		arr[l] = arr[r];
		arr[r] = temp;
	}
	quicksort(left, l - 1);
	quicksort(r + 1, right);
}

int main() {
	int n;
	scanf("%d", &n);
	arr = new int[n];
	for (int i = 0; i < n; i++)
		scanf("%d", &arr[i]);
	quicksort(0, n - 1);
	for (int i = 0; i < n; i++)
		printf("%d\n", arr[i]);
	return 0;
}